src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -390,19 +392,82 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
                         _includeCtes = new List<string>();
                     }
 
+                    _includeCtes.Add(TableExpressionName(_tableExpressionCounter));
+                    break;
+                case TableExpressionKind.RevInclude:
+                    var revIncludeExpression = (RevIncludeExpression)tableExpression.NormalizedPredicate;
+
+                    StringBuilder.Append("SELECT ");
+                    StringBuilder.Append(VLatest.Resource.ResourceSurrogateId, referenceSourceTableAlias).AppendLine(" AS Sid1 ");
+                    StringBuilder.Append(", (CASE WHEN CountSource > " + MaxIncludedItems + " THEN ")
+                    .Append(VLatest.Resource.ResourceSurrogateId, referenceTargetResourceTableAlias)
+                    .Append(" ELSE NULL END) IsPartialTargetSid1 ");
+
+                    StringBuilder.AppendLine("FROM ").Append(VLatest.Resource).Append(' ').AppendLine(referenceTargetResourceTableAlias);
+                    StringBuilder.AppendLine("CROSS APPLY -- we want the top to work on the (rev)included resource records and not on the overall resultset, hence can't use a simple join.");
+                    StringBuilder.AppendLine("( ");
+
+                    StringBuilder.AppendLine("SELECT DISTINCT TOP (" + MaxIncludedItems + ") ResourceSurrogateId")
+                    .AppendLine(", count(*) over() as CountSource -- count all records before the TOP")
+                    .AppendLine("FROM ").Append(VLatest.ReferenceSearchParam).Append(' ').AppendLine(referenceSourceTableAlias);
+
+                    using (var delimited = StringBuilder.BeginDelimitedWhereClause())
+                    {
+                        delimited.BeginDelimitedElement().Append(VLatest.ReferenceSearchParam.ReferenceResourceTypeId, referenceSourceTableAlias)
+                            .Append(" = ").Append(VLatest.Resource.ResourceTypeId, referenceTargetResourceTableAlias);
+
+                        delimited.BeginDelimitedElement().Append(VLatest.ReferenceSearchParam.ReferenceResourceId, referenceSourceTableAlias)
+                            .Append(" = ").Append(VLatest.Resource.ResourceId, referenceTargetResourceTableAlias);
+
+                        if (!revIncludeExpression.WildCard)
+                        {
+                            delimited.BeginDelimitedElement().Append(VLatest.ReferenceSearchParam.SearchParamId, referenceSourceTableAlias)
+                                .Append(" = ").Append(Parameters.AddParameter(VLatest.ReferenceSearchParam.SearchParamId, Model.GetSearchParamId(revIncludeExpression.ReferenceSearchParameter.Url)));
+                        }
+
+                        AppendHistoryClause(delimited, referenceSourceTableAlias);
+                    }
+
+                    StringBuilder.AppendLine(") AS ").Append(referenceSourceTableAlias).Append(" ");
```

Comment:
I think this new query might also need to include an order by - in other places you see that we always return the top-x records order by ResourceSurrogateId (usually ASC but there's logic to switch direction). Although this isn't exactly the same case (i.e. probably the request can't ask this sort to be different), we might want to keep it consistent.
Anyway, an observation I had...

Reply:
Fixed